dash_core_components/_imports_.py

```diff
@@ -36,14 +35,13 @@
     "Loading",
     "Location",
     "LogoutButton",
-    "Markdown",
+    "markdown",
```

Comment:
@alexcjohnson This isn't very robust, because it must be manually edited after a build. But I'm not sure how to change the auto-generated `_imports_.py`. I figured it might require a change to `dash-generate-components`. However, it may not be needed. The dependency, `highlight.js`, is much smaller than the now-unneeded `react-syntax-highlighter`.

Reply:
Hmm yeah that's not great... also as it is, `markdown` will always be imported anyway so we haven't saved its bytes (aside from what we've already saved by swapping `highlight.js` in for `react-syntax-highlighter`). `__init__.py` isn't auto-generated, right? Seems like we can:
- leave `_imports_.py` and `Markdown.py` where they go automatically
- in `__init__.py`, filter out `Markdown` by changing the imports like:
```py
import ._imports_ as _imports
__all__ = []
for _name in dir(_imports):
    if _name != 'Markdown' and not _name.startswith('_'):
        locals()[_name] = _imports[_name]
        __all__.append(_name)
```
- add a `markdown.py` that looks exactly like your `markdown.__init__.py`

Then, to use `markdown.py` and include its dist files, you have to do `from dash_core_components.markdown import Markdown`

One problem with this is if people make a mistake and import `dash_core_components.Markdown` it won't throw an error in python, it'll just fail to render. I suppose we could at least make the `if (!window.hljs)` block throw a clear error telling you how to import the component correctly? (same for `react-markdown` if we manage to pull that out too)

Reply:
If people do `dcc.markdown.Markdown` after the app's start (e.g. inside a callback), then the dist won't be included right?

Reply:
Right, that wouldn't work either. OK, so maybe it's premature to try and break this stuff out, too many ways to break it and I don't see a clean way around it. Let's file this under "should be solved by lazy loading" and put `Markdown` back on par with everything else - ie a regular top-level import whose js deps are always available.

Reply:
👍

Reply:
OK, I'll do this in a more conventional way and include `react-markdown` and `highlight.js` in the dcc `_js_dist`. If I do this, it will only add about 20kB. Before, with `SyntaxHighlighter`, it added 188kB from `react-syntax-highlight`, and 740kB (!) from `highlight.js`. Installing `highlight.js` via NPM is ~[533kB](https://bundlephobia.com/result?p=highlightjs@9.12.0) :scream_cat:.

The default build of `highlight.js` is 50kB with support for "24 commonly used languages". My custom build is only 12kB; IIRC I'm only including Python, CSS, Javascript and Markdown support. You can easily create a customized `highlight.js` [here](https://highlightjs.org/download/#cdns).

I'll suggest a `highlight.js` with the following languages supported (20kB minified):

- CSS
- HTTP
- JavaScript
- Python
- Bash
- JSON
- Markdown
- HTML, XML
- SQL (?)
- "Shell Session" (``` shell ... ```)

Obviously we can always add more later without introducing any breaking changes. Any languages I'm missing?

Reply:
R

Reply:
Oh that's awesome! I'd just add TypeScript, and yes let's keep SQL.

At some point we can make some utilities to pick from a few bundle options for both highlight.js and plotly.js - including perhaps an option to turn both of these off entirely and let you put your own in `assets`.

Reply:
And in fact, if `react-markdown` is a single thing and isn't expected to be configurable, I'd leave it in the main dcc bundle, keep it off `window`